File: Scala/src/main/scala/TextbookTrack/Chapter09/BA9H.scala
package TextbookTrack.Chapter09

object BA9H {
  import scala.annotation.tailrec
  import TextbookTrack.RichNumber._

  private def sortSingleCharacters(text: String): (Vector[Int], Vector[Int]) = {
    val indices: Map[Char, IndexedSeq[Int]] = text.zipWithIndex.groupMap{ case (char, _) => char }{ case (_, ix) => ix }
    val uniqueChars: Vector[Char] = indices.keys.toVector.sorted
    val charIds: Map[Char, Int] = uniqueChars.zipWithIndex.toMap
    val order: Vector[Int] = uniqueChars.flatMap(indices(_))
    val classes: Vector[Int] = text.map(charIds(_)).toVector
    (order, classes)
  }

  private def countCumulativeClassSizes(classes: Vector[Int], n: Int): Array[Int] = {
    val cumulativeCounts: Array[Int] = Array.fill(n)(0)
    val classSizes: Map[Int, Int] = classes.groupMapReduce(identity)(_ => 1)(_ + _)
    classSizes.foreach{ case (classId, classSize) => cumulativeCounts(classId) = classSize }
    (1 until n).foreach(ix => cumulativeCounts(ix) += cumulativeCounts(ix - 1))
    cumulativeCounts
  }

  private def sortDoubledShifts(order: Vector[Int], classes: Vector[Int], n: Int, cyclicShiftSize: Int): Vector[Int] = {
    val counts: Array[Int] = countCumulativeClassSizes(classes, n)
    val newOrder: Array[Int] = Array.fill(n)(0)
    for { ix <- (n - 1) to 0 by -1 } {
      val start: Int = (order(ix) - cyclicShiftSize) mod n
      val classId: Int = classes(start)
      counts(classId) -= 1
      newOrder(counts(classId)) = start
    }
    newOrder.toVector
  }

  private def updateClasses(order: Vector[Int], classes: Vector[Int], cyclicShiftSize: Int, n: Int): Vector[Int] = {
    val newClasses: Array[Int] = Array.fill(n)(0)
    newClasses(order(0)) = 0
    val orders: Iterator[(Int, Int)] = order.sliding(2).collect{ case Vector(a, b) => (a, b) }

    @tailrec
    def loop(label: Int): Unit = {
      if (orders.isEmpty) ()
      else {
        val (previous, current): (Int, Int) = orders.next()
        val mid: Int = (current + cyclicShiftSize) % n
        val midPrevious: Int = (previous + cyclicShiftSize) % n
        val nextLabel: Int =
          if (classes(current) != classes(previous) || classes(mid) != classes(midPrevious)) label + 1 else label
        newClasses(current) = nextLabel
        loop(nextLabel)
      }
    }

    loop(0)
    newClasses.toVector
  }

  private def computeSuffixArray(text: String): Vector[Int] = {
    val n: Int = text.length
    val (order, classes): (Vector[Int], Vector[Int]) = sortSingleCharacters(text)

    @tailrec
    def loop(cyclicShiftSize: Int, cyclicShiftOrders: Vector[Int], equivalenceClasses: Vector[Int]): Vector[Int] = {
      if (cyclicShiftSize >= n) cyclicShiftOrders
      else {
        val doubleShiftOrders: Vector[Int] =
          sortDoubledShifts(cyclicShiftOrders, equivalenceClasses, n, cyclicShiftSize)
        val updatedClasses: Vector[Int] = updateClasses(doubleShiftOrders, equivalenceClasses, cyclicShiftSize, n)
        loop(2 * cyclicShiftSize, doubleShiftOrders, updatedClasses)
      }
    }

    loop(1, order, classes)
  }

  private def patternMatchingWithSuffixArray(suffixArray: Vector[Int],
                                             text: String,
                                             n: Int,
                                             pattern: String): List[Int] = {
    @tailrec
    def findMinIndex(minIndex: Int, maxIndex: Int): Int = {
      if (minIndex >= maxIndex) minIndex
      else {
        val middleIndex: Int = (minIndex + maxIndex) / 2
        if (pattern > text.substring(suffixArray(middleIndex))) findMinIndex(middleIndex + 1, maxIndex)
        else findMinIndex(minIndex, middleIndex)
      }
    }

    @tailrec
    def findMaxIndex(minIndex: Int, maxIndex: Int): Int = {
      if (minIndex >= maxIndex) maxIndex
      else {
        val middleIndex: Int = (minIndex + maxIndex) / 2
        val textSlice = text.substring(suffixArray(middleIndex))
        if (textSlice.startsWith(pattern)) findMaxIndex(middleIndex + 1, maxIndex)
        else if (pattern < textSlice) findMaxIndex(minIndex, middleIndex)
        else findMaxIndex(middleIndex + 1, maxIndex)
      }
    }

    val minIndex: Int = findMinIndex(0, n)
    val maxIndex: Int = findMaxIndex(minIndex, n)
    if (minIndex > maxIndex) Nil
    else (minIndex until maxIndex).map(suffixArray).toList
  }

  def multiplePatternMatching(text: String, patterns: List[String]): Set[Int] = {
    val suffixArray: Vector[Int] = computeSuffixArray(text)
    val n: Int = text.length
    patterns.foldLeft(Set.empty[Int]){
      (matchedIndices, pattern) => matchedIndices ++ patternMatchingWithSuffixArray(suffixArray, text, n, pattern)
    }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val text: String = reader.next() + '$'
    val patterns: List[String] = reader.toList
    val result: Set[Int] = multiplePatternMatching(text, patterns)
    println(result.mkString(", "))
  }
}

File: Python/textbook_track/chapter09/ba9h.py
# Pattern Matching with the Suffix Array
import itertools as it
import sys

from collections import Counter, defaultdict


def read_patterns(reader):
    return list(map(lambda s: s.rstrip(), reader))


def sort_single_characters(text):
    indices = defaultdict(list)
    for ix, char in enumerate(text):
        indices[char].append(ix)

    unique_chars = sorted(indices.keys())
    char_indices = (indices[char] for char in unique_chars)
    return tuple(it.chain.from_iterable(char_indices))


def compute_character_classes(text, order, n):
    classes = [0] * n
    label = 0
    classes[order[0]] = label
    for o1, o2 in zip(order, order[1:]):
        if text[o1] != text[o2]:
            label += 1
        classes[o2] = label
    return tuple(classes)


def count_cumulative_class_sizes(classes, n):
    counts = [0] * n
    class_sizes = Counter(classes)
    for class_id, class_size in class_sizes.items():
        counts[class_id] = class_size

    for ix in range(1, n):
        counts[ix] += counts[ix-1]

    return counts


def sort_doubled_shifts(order, classes, n, cyclic_shift_size):
    counts = count_cumulative_class_sizes(classes, n)
    new_order = [0] * n
    for ix in range(n-1, -1, -1):
        start = (order[ix] - cyclic_shift_size) % n
        class_id = classes[start]
        counts[class_id] -= 1
        new_order[counts[class_id]] = start
    return tuple(new_order)


def update_classes(order, classes, cyclic_shift_size, n):
    new_classes = [0] * n
    label = 0
    new_classes[order[0]] = label
    for current, previous in zip(order[1:], order):
        mid = (current + cyclic_shift_size) % n
        mid_previous = (previous + cyclic_shift_size) % n
        if classes[current] != classes[previous] or classes[mid] != classes[mid_previous]:
            label += 1
        new_classes[current] = label
    return tuple(new_classes)


def compute_suffix_array(text):
    n = len(text)
    order = sort_single_characters(text)
    classes = compute_character_classes(text, order, n)
    cyclic_shift_size = 1
    while cyclic_shift_size < n:
        order = sort_doubled_shifts(order, classes, n, cyclic_shift_size)
        classes = update_classes(order, classes, cyclic_shift_size, n)
        cyclic_shift_size *= 2
    return order


def pattern_matching_with_suffix_array(suffix_array, text, n, pattern):
    min_index = 0
    max_index = n
    while min_index < max_index:
        middle_index = (min_index + max_index) // 2
        if pattern > text[suffix_array[middle_index]:]:
            min_index = middle_index + 1
        else:
            max_index = middle_index
    first = min_index

    max_index = n
    while min_index < max_index:
        middle_index = (min_index + max_index) // 2
        suffix = text[suffix_array[middle_index]:]
        if suffix.startswith(pattern):
            min_index = middle_index + 1
        elif pattern < suffix:
            max_index = middle_index
        else:
            min_index = middle_index + 1
    last = max_index

    if first > last:
        return []

    return [suffix_array[index] for index in range(first, last)]


def multiple_pattern_matching(text, patterns):
    suffix_array = compute_suffix_array(text)
    n = len(text)
    matched_indices = set()
    for pattern in patterns:
        indices = pattern_matching_with_suffix_array(suffix_array, text, n, pattern)
        matched_indices.update(indices)
    return matched_indices


def main():
    reader = sys.stdin
    text = next(reader).rstrip() + '$'
    patterns = read_patterns(reader)
    result = multiple_pattern_matching(text, patterns)
    print(' '.join(map(str, result)))


if __name__ == "__main__":
    main()
